### DISCUSSION TITLE
/hyndsight/tscharacteristics/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#111] [Announcements]
Measuring time series characteristics 
 https://robjhyndman.com/hyndsight/tscharacteristics/

**Comment by robjhyndman** [comment]
```header
Originally posted by fmark on 2012-05-02 09:36:25
````
Thanks for posting your code Rob. This is very similar to a project I
envisaged a couple of years ago, before I had the skills to put the
ideas into practice. I\'ll let you know if I find the time to publish a
paper based on the code.

**Comment by robjhyndman** [comment]
```header
Originally posted by Guest on 2012-05-03 00:29:15
````
If there a way to use the forcast function out of the measures out put
function ?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-05-03 00:30:06
````
I don\'t understand what you mean. What do you want to forecast?

**Reply by robjhyndman** [reply]
```header
Originally posted by Guest on 2012-05-03 00:46:06
````
I was looking at one of the many forecast package example:

    fit \&lt;- Arima(WWWusage,order=c(3,1,0))\
    plot(forecast(fit,h=20))Is it possible to imagine doing something
similar to:\
   fit2 \&lt;- measures(WWWusage)   plot(forecast(fit2,h=20))?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-05-03 00:54:28
````
forecast(fit) produces forecasts from an ARIMA model. How can you
forecast from the measures? These are just summary statistics of the
time series.

**Reply by robjhyndman** [reply]
```header
Originally posted by Guest on 2012-05-03 01:20:56
````
Sorry, my question may not have any sense. From the measures function
output, timeseries can be clustered. Linking (how?) clusters with
appropriate models, a forecast can then be performed. For instance, a
cluster with strong trend and seasonnality components could be fitted
with fit \&lt;- tslm(y \~ trend + season) and then forcasted.

**Reply by robjhyndman** [reply]
```header
Originally posted by Pavan Nagula on 2018-12-06 04:34:40
````
Hi Rob, Can we include this code as part of forecast package? given a
timeseries data input, measures function will emit summary statistics of
timeseries data. This will help to include these stats as features in
classification or clustering process.\
Thanks,\
Pavan Nagula

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-12-06 05:07:06
````
My tsfeatures package does the features calculation:
[https://github.com/robjhynd\...](https://github.com/robjhyndman/tsfeatures)

**Reply by robjhyndman** [reply]
```header
Originally posted by Pavan Nagula on 2018-12-06 05:10:38
````
Wow, I didn\'t know there is a package like that.. Thanks Rob for your
quick reply!!!

**Comment by robjhyndman** [comment]
```header
Originally posted by Abhijith B on 2012-05-03 08:45:38
````
How can we automate a timeseries procedure to find the orders of ARIMA
in R (not with auto.arima()).\
Suppose if there are more than one time series data are available.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-05-03 22:25:32
````
What\'s wrong with using auto.arima()? If you have more than one time
series, use auto.arima() on each one.

**Comment by robjhyndman** [comment]
```header
Originally posted by Paul Prew on 2012-05-03 16:24:26
````
Hello, This looked very promising for aproject I\'m working on.  I tried
to run the code, but got an error message.  Maybe it\'s becasue I never
specified the (a,b) arguments?  I ran the code in blocks as grouped in
your post.  I know R a bit, mostly interact using RCommander.  I\'ll be
the first to admit my data is strange, but it\'s actual pricing history
for a raw material.

\&gt; srn.dat\$srn172163.100 \&lt;-
ts(srn.dat\$srn172163.100,start=c(2008,2),freq=1)\
 \
\&gt; x\&lt;-srn.dat\$srn172163.100\
 \
\&gt; summary(x)\
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.\
 0.6250  0.6350  0.6850  0.7089  0.7550  0.8810\
 \
\&gt; dput(x)\
structure(c(0.705, 0.705, 0.7077561766, 0.725, 0.725, 0.789008964,\
0.8664955774, 0.868, 0.881, 0.8706588748, 0.8199582478, 0.725,\
0.675, 0.625, 0.625, 0.625, 0.625, 0.625, 0.625, 0.625, 0.625,\
0.635, 0.635, 0.635, 0.635, 0.635, 0.7920070187, 0.8376571637,\
0.755, 0.755, 0.685, 0.685, 0.685, 0.685, 0.685, 0.7180954833\
), .Tsp = c(2009, 2044, 1), class = &quot;ts&quot;)\
 \
\&gt; f1 \&lt;- function(x,a,b)\
...\
return(f1eax)}\
\*\*\* no messages in log, so assume no new is good news and f1 executed
OK despite no specification of the (a,b) arguments.  Probably a bad
assumption:

\&gt;summary(f1eax)\
ERROR:  object \'f1eax\' not found

measures \&lt;- function(x)\
  return(m)}\&gt;m\
\[7\] ERROR:  object \'m\' not found

\&gt; ls() \[1\] &quot;decomp&quot;         &quot;f1&quot;             &quot;f2&quot;            
&quot;find.freq&quot;      \[5\] &quot;snr.dat&quot;        &quot;snr.datsrn.dmx&quot;
&quot;srn.dat&quot;        &quot;srn.dmx&quot;        \[9\] &quot;srn.dmy&quot;       
&quot;srn.tslars&quot;     &quot;x&quot;

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-05-03 22:27:07
````
Use measures(x)

**Comment by robjhyndman** [comment]
```header
Originally posted by Paul Prew on 2012-05-04 18:11:40
````
Thank you, Rob.  Worked perfectly.  Next month I start an R programming
class, maybe I can avoid these elementary errors in the future.

I see there\'s auto.arima, and this automated measures script \-- is
there any automated procedure for transfer functions?  We searches My
project involves forecasting raw material cost from price indices of
certain commodities.  There\'s a couple thousand raw materials, and I
thought the time series clustering might shed some light.  But an
automated transfer seems even more valuable.

Thanks again for making this code available.  regards, Paul

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-05-05 13:41:48
````
Automated transfer functions would be cool, but it isn\'t available in R
at this stage. The commercial package Autobox will do it, I think.

**Comment by robjhyndman** [comment]
```header
Originally posted by Umer Zeeshan Ijaz on 2012-05-05 17:54:52
````
Thanks for sharing the scripts Rob. They are quite helpful! Installing
forecast package on Ubuntu (11.04 Natty) was a big pain though as I
found out that I can only install &quot;forecast&quot; when the current version
of R \&gt;=2.14. In my case, I was running 2.12 and so I had to not only
install the newer version of R but also update the GCC to version 4.6.
In my case, one of the dependencies, &quot;quadprog&quot; wasn\'t compiling with
the current version (I think it was 4.5) of GCC.

Having fixed all these problems, I am writing what I did to resolve
these issues so that those of us who have debian distros and are facing
the same problem can save some hassle.

\###### If GCC is a problem \######

The following steps are more or less similar to what is mentioned in the
following blogpost.
[http://buildall.wordpress.c\...](http://buildall.wordpress.com/2011/04/20/installing-gcc-4-6-in-the-ubuntu-10-10/ &quot;http://buildall.wordpress.com/2011/04/20/installing-gcc-4-6-in-the-ubuntu-10-10/&quot;)

Following this, I got the GCC from
[http://gcc.parentingamerica\...](http://gcc.parentingamerica.com/snapshots/LATEST-4.6/ &quot;http://gcc.parentingamerica.com/snapshots/LATEST-4.6/&quot;)     
(gcc-4.6-20120504.tar.bz2)

Untar it:\
tar xjvf gcc-4.6-20110401.tar.bz2

This version of gcc requires the following libraries: gmp-4.3.2,
mpfr-2.4.2, mpc-0.8.1. To get these libraries, run the following script
in the untarred folder :\
gcc-4.6-20120427/contrib/download_prerequisites

Go to each folders for the above libraries that are created in
gcc-4.6-20120427/contrib and install using either:\
./configure\
make install\
or\
./install-sh

Now we are ready to compile gcc. Come out of gcc-4.6-20120427 and make a
folder called build:\
mkdir build

Now you should have two folders in the current directory i.e.\
\&lt;current path=&quot;&quot;\&gt;/build\
\&lt;current path=&quot;&quot;\&gt;/gcc-4.6-20110401

Go inside the build folder and run the following\
../gcc-4.6-20110401/configure \\\
disable-checking \\\
enable-languages=c,c++ \\\
enable-multiarch \\\
enable-shared \\\
enable-threads=posix \\\
program-suffix=-4.6 \\\
with-gmp=/usr/local/lib \\\
with-mpc=/usr/lib \\\
with-mpfr=/usr/lib \\\
without-included-gettext \\\
with-system-zlib \\\
with-tune=generic\
make\
sudo make install

Update your Ubuntu to use the new GCC:\
cd /usr/bin\
sudo rm g++\
sudo ln -s /usr/local/bin/g++-4.6 g++\
sudo rm gcc\
sudo ln -s /usr/local/bin/gcc-4.6 gcc

\##### If using an older version of R \#####

Update R as:\
sudo apt-get install r-base\
sudo apt-get install r-recommended

Now inside R check if you have the latest version\
R.Version()

Install forecast package now\
install.packages(&quot;forecast&quot;)

If you are accessing internet through your proxy server, set the
following environmental variable in R before you run the above command\
Sys.setenv(http_proxy=&quot;\&lt;proxyaddr:port\&gt;&quot;)

I was still unable to get quadprog to install. It required libgfortran
to be in /usr/lib/ whereas in my Lenovo thinkpad\'s ubuntu distro, by
default it was located at /usr/lib/x86_64-linux-gnu/\
I used the following command to find it:\
ldconfig -p \| grep -i &quot;libgfortran&quot;

I then made a soft symbolic link to it in /usr/lib using\
ln -s /usr/lib/x86_64-linux-gnu/libgfortran.so.3.0.0
/usr/lib/[libgfortran.so](http://libgfortran.so &quot;libgfortran.so&quot;)

\##### If RKward stopped working after upgrading R \#####

I use RKward
([http://rkward.sourceforge.n\...](http://rkward.sourceforge.net/) &quot;http://rkward.sourceforge.net/)&quot;)
as a GUI editor to program in R. After updating R, it stopped working
properly. It gave the message &quot;The \'rkward\' R-library either could
not be loaded at all, or not in the correct version&quot; and then crashed
as soon as I ran any R commands. Here is the fix for it:

Remove RKward using any of the following commands:\
sudo apt-get purge rkward / sudo apt-get autoremove rkward / sudo
apt-get remove rkward

Make sure that the library is not deinstalled/deselected in dpkg\
dpkg \--get-selection \| grep -i &quot;rkward&quot;

If found then purge it using\
dpkg -P rkward

Make sure you remove the following directory as well\
rm -r \~/.rkward

Install rkward using this ppa\
sudo add-apt-repository ppa:rkward-devel/rkward-stable-cran\
sudo apt-get update

Regards,\
Umer
([http://userweb.eng.gla.ac.u\...](http://userweb.eng.gla.ac.uk/umer.ijaz) &quot;http://userweb.eng.gla.ac.uk/umer.ijaz)&quot;)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-05-05 23:50:05
````
The latest versions of forecast now require R2.14 so they can use code
compiling and RcppArmadillo to speed things up.

**Comment by robjhyndman** [comment]
```header
Originally posted by Yanfei Kang on 2012-05-06 10:51:27
````
Hi Rob, 

Thanks for making the codes public. It works perfectly except the
kurtosis and Dc kurtosis. You may find in your example on Australian
monthly gas production, both the kurtosis and Dc kurtosis are 1.0000,
which is kind of wired. By trying the codes on several other time
series, I think the reason is as follows:In 

 # Kurtosis

  k \&lt;- mean((tadj.x-xbar)\^4,na.rm=TRUE)/s\^4

  fk \&lt;- f1(k,2.273,11567),

the \$s\$ is actually the value taken from:

  # Skewness

  s \&lt;- abs(mean((tadj.x-xbar)\^3,na.rm=TRUE)/s\^3)

  fs \&lt;- f1(s,1.510,5.993)\
but not the wanted \$s\$ value from 

  s \&lt;- sd(x,na.rm=TRUE).

In your example, the skewness value before transformation should be very
small, which causes \$k\$ to be infinity. That\'s why the transformed
kurtosis \$fk\$ is 1.0000. Also for Dc kurtosis.

Thanks again for making the codes available.Regards,yanfei

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-05-06 10:55:52
````
Oops. Good catch. That error was introduced when I modified the code for
the blog post \-- it wasn\'t in the code we used for the paper. Anyway,
now fixed. Thanks.

**Comment by robjhyndman** [comment]
```header
Originally posted by Scott Locklin on 2012-12-11 04:57:16
````
Thanks kindly for that source code. While I love the SOM\'s, I couldn\'t
get it to work that way. It does work pretty well with proxy::dist and
stats::hclust + stats::cutree\
This is a great trick: I used something similar to it + knn to detect
regimes in subsegments of very large time series on another problem.

Your version of the technique is better for shorter/lower-frequency
timeseries. The lack of adequate clustering algorithms on CRAN cries out
for a remedy, even if it is a pastiche of many packages. Maybe it\'s
time for me to write one\...

**Comment by robjhyndman** [comment]
```header
Originally posted by John Williams on 2013-08-22 06:47:22
````
Thanks heaps Rob. Just a quick note: on my system (R 3.0.1) I had to add
a couple of requires: tseries and fracdiff. But that\'s trivial. Thanks
again! :-)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-22 06:55:56
````
Thanks. When I wrote this, the forecast package automatically loaded
those packages. More recent versions of forecast now only import the
functions needed, so there is now a need to add those requires.

**Comment by robjhyndman** [comment]
```header
Originally posted by SriV on 2013-09-27 16:06:58
````
Dear Rob,\
You code is very informative. Thanks. Regarding the \'normalization\' ,
particularly, the function &quot;f1&quot; which converts any \[0, Infininity\]
to \[0, 1\], can you please explain that function a bit more? I am
unable to understand it. Also, what are those three arguments to the
function - x, a and b? I looked at its usage further down, but, still
could not understand.

Thanks,\
Vats.

**Comment by robjhyndman** [comment]
```header
Originally posted by Baman on 2014-01-18 19:06:50
````
Thank you for publishing these helpful codes, I\'m trying to use the
measure() function for my data set, but that gives me an error for some
of them

\&gt; x

Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec

2001 9 20 13 33 25 9 5 34 20 11 7 8

2002 5 4 10 46 71 32 10 15 4 18 7 5

2003 2 5 5 9 26 18 16 16 6 8 1 13

2004 5 9 7 32 33 10 24 16 42 33 18 38

2005 17 13 2 10 12 46 23 37 11 20 9 23

2006 33 14 31 17 39 20 18 31 26 48 30 53

2007 10 32 27 27 29 46 33 22 32 30 29 51

2008 44 20 42 35 42 14 30 26 62 36 34 29

2009 36 37 25 42 32 48 20

measures(x)

Error in stl(x, s.window = &quot;periodic&quot;, na.action = na.contiguous) :

only univariate series are allowed

Could you please give me any suggestion to correct it?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-01-19 02:36:58
````
It looks like your data is in a dataframe or a matrix object, rather
than a ts object. It must be a univariate ts object.

**Comment by robjhyndman** [comment]
```header
Originally posted by Geraldine  on 2014-02-16 14:49:52
````
Dear Rob\
Thank you so much for posting your code. It is exactly what I was
looking for.\
I\'m currently working on implementing the second block of your code. Am
I wrong that it doesn\'t work with time series \&lt; 10 periods? It seems
like the gam function can not handle these short time series. I replaced
it with an lm function and that seems to work ok. Do you have any idea
what else I can do to fix that problem? Around 40% of my time series
have less than 10 periods.

Best Geraldine

**Comment by robjhyndman** [comment]
```header
Originally posted by A on 2015-03-16 07:41:27
````
Hi Rob,

Thanks for posting this r example its exactly what i was looking for.
Along the same lines as above could you use something similar for
pattern detection in transactions. i have a sparse transaction data set
and would like to generate some summary stats like periodicity, \$
amount that is periodic and others \... any thoughts?

**Comment by robjhyndman** [comment]
```header
Originally posted by Ch on 2015-05-02 15:08:50
````
Excellent blog I must say!

The find.freq function works brilliantly. On the daily data set I am
using, it correctly worked out the frequency to be 7.

When I tried it on only the week days, it mentioned the frequency is 23,
which is remarkably close to 21.42857=29.6\*5/7 which is the average
number of work days in a month. (Or conversely 23\*7/5 is 32.)

Is there a function that gives the list of all the seasonal.periods for
use in msts(\...)?

I experimented with a hunch of taking the first period, averaging by
that and then finding that period, etc.

    find.freq.all=function(x){  
      f=find.freq(x);
      freqs=c(f);  
      while(f&gt;1){
    start=1; #also try start=f;
    x=period.apply(x,seq(start,length(x),f),mean); 
        f=find.freq(x);
        freqs=c(freqs,f);
      }
    if(length(freqs)==1){ return(freqs); }
      for(i in 2:length(freqs)){
        freqs[i]=freqs[i]*freqs[i-1];
      }
      freqs[1:(length(freqs)-1)];
    }
    find.freq.all(dailyts)  #using daily data

\
The above gives (7,28) or (7,35) depending on if the seq starts with 1
or f. (See comment above.)

Which would imply that the seasonal periods for msts should be (7,28) or
(7,35).

The logic appears sensitive to initial conditions given the sensitivity
of the algorithm parameters. The mean of 28 and 35 is 31.5 which is
close to the average length of a month.

I suspect I reinvented the wheel, what is the name of this algorithm?

Any thoughts or comments?

P.S. Later, I ran the above code in trying all starts of 1:7 and I got
35,35,28,28,28,28,28 for the second period. The average works out to 30.
Interesting\...

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-05-03 01:07:13
````
I think you would be better off trying to remove the seasonality of the
first frequency before looking for the next one. I\'m not sure what your
period.apply() function does here. You could use stl to estimate the
first seasonal pattern, remove it, and then use find.freq() to look at
what\'s left.

**Reply by robjhyndman** [reply]
```header
Originally posted by Ch on 2015-05-03 03:29:12
````
The period.apply in the above code determines weekly daily average.\
Changing that line to x=period.apply(x,seq(start,length(x),f),sum);
computes weekly totals and gives the same results.

When I tried stl() it gives &quot;series is not periodic or has less than
two periods.&quot;

However, I know that the first of the month (that is a weekday) there is
additional web traffic over and above the usual. Thus it determines that
there is an additional spike every 4th or 5th week.\
Overall there is a growth pattern. Doesn\'t that beginning of the month
activity count as a seasonal pattern?

**Comment by robjhyndman** [comment]
```header
Originally posted by sasuke on 2015-10-27 08:58:12
````
should the result of the function be used as frequency for my time
series\
in my case I have daily data with an observation every 15 min,total of
96 per day, of 3 months worth of data from last year and this year and
nothing in between

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-10-27 22:04:07
````
I assume you mean the find.freq() function. If you know the frequency,
you don\'t need to estimate it. In your case, you have potential
seasonal periods of 96, 96\*7 and 96\*365. Or use all three with an msts
object.

**Comment by robjhyndman** [comment]
```header
Originally posted by Clemens on 2016-05-03 16:13:10
````
Thanks a lot for the code. It works very well. However, I believe that
the line

nextmax \&lt;- j\[1\] + which.max(spec\$spec\[j\[1\]:500\])

should be

nextmax \&lt;- j\[1\] + which.max(spec\$spec\[j\[1\]:500\]) - 1

to get the index right. Also, it might make the code easier to
understand if the number 500 is replaced with a variable (i.e. nfreq)
and to also include this variable in the call to
[spec.ar](http://spec.ar &quot;spec.ar&quot;)

spec \&lt;-
[spec.ar](http://spec.ar &quot;spec.ar&quot;)(c(na.contiguous(x)),plot=FALSE,n.freq=nfreq)

Thanks again for the great code.\
Regards, Clemens

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-05-03 22:36:24
````
Thanks. Yes, I was off by one there although it makes almost no
difference as the resulting frequency is rounded to an integer. This
function has been improved slightly and is now available as
findfrequency() in the forecast package.

**Comment by robjhyndman** [comment]
```header
Originally posted by Kathrine Makhnatch on 2016-07-01 08:40:09
````
Thanks for sharing the script. It\'s very helpful!\
But in the function maping \[0,infinity) to \[0,1\] it will be better to
replace eax == Inf by is.infinite(eax).

**Reply by robjhyndman** [reply]
```header
Originally posted by Yuqi Wang on 2018-04-12 19:11:27
````
Thank you! This helps fix my code when i use the function!!

**Comment by robjhyndman** [comment]
```header
Originally posted by Rakesh Tripathi on 2016-08-04 12:32:34
````
How to judge the accuracy of time series decomposition provided by R?

Because at different level of data aggregation and seasonality different
repetitive patterns appear in seasonality index.

How to know that what is the best suited seasonality and granularity of
data on which the decomposition is reliable?

I am trying to find signal to noise ration where signal is considered as
trend + seasonality and noise is reminder component.

Is there any theoretical backing through which we can be confident of
that at least we can rely on this decomposition for underlying patterns.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-08-04 23:01:32
````
This is not a help site. Please ask general questions on
[crossvalidated.com](http://crossvalidated.com &quot;crossvalidated.com&quot;)

**Comment by robjhyndman** [comment]
```header
Originally posted by Fabrizio Maccallini on 2016-12-06 09:51:20
````
Thank you, it\'s people with your attitude that will make our future
great! I have a question: in the first block of code when you compute
the spectrum through the
[spec.ar](http://spec.ar &quot;spec.ar&quot;) function,
is there a reason the method should be Yule-Walker (default) rather than
Burg?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2016-12-06 11:26:16
````
No, either should be ok.

**Comment by robjhyndman** [comment]
```header
Originally posted by Uttara Ananthakrishnan on 2017-04-03 23:48:31
````
Thank you so much for this code. I was wondering if there was a reason
to set the lag at 10 for Box Test (For the Q-Statistic). Is it standard
practice in time series data or was it data specific?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-04-04 00:32:27
````
See
[http://robjhyndman.com/hynd\...](http://robjhyndman.com/hyndsight/ljung-box-test/ &quot;http://robjhyndman.com/hyndsight/ljung-box-test/&quot;)

**Reply by robjhyndman** [reply]
```header
Originally posted by Uttara Ananthakrishnan on 2017-04-04 01:11:00
````
Thanks! It is super helpful

**Comment by robjhyndman** [comment]
```header
Originally posted by Rohit on 2017-06-28 12:58:14
````
Thanks for such an informative post!

Quick question on the trend &amp; seasonal measures. Why are you adding the
mean of trend to the adjusted data?\
adj.x \&lt;- decomp.x\[\[&quot;x&quot;\]\] - fits + mean(decomp.x\[\[&quot;trend&quot;\]\],
na.rm=TRUE)

Also it will be great if you can explain the following calculation :\
max(0,min(1,1-v.adj/var(deseason,na.rm=TRUE)))

Is the below interpretation correct?\
Lower the ratio of \[variance of adjusted data\] to \[variance of
deseason data\], higher variance in trend

Appreciate your help! Thanks again

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-29 06:44:51
````
1\. Mean added so the adjusted data has same level as original data.
Otherwise it would have mean close to 0.\
2. Let r = 1 minus the ratio of variances (as explained in the paper).
If r\&lt;0, it returns 0. If r\&gt;1, it returns 1. Otherwise it returns r.

**Reply by robjhyndman** [reply]
```header
Originally posted by Rohit on 2017-06-29 08:44:30
````
Apologies for not being clear in my question. I understand the math you
mentioned in point 2. I am struggling to understand the reason for this
math to be a measure of trend (strength)? Thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-06-29 08:56:37
````
x=Trend+Seasonal+Remainder.

Adjusted = x-Trend-Seasonal=Remainder.

Deseason = x-Seasonal = Trend+Remainder

So V(adjusted) includes only remainder variation while V(Deseason)
includes both remainder variation and trend variation. Hence
V(adjusted)/V(Deseason) will be small if trend variation is high, but
large (close to 1) if there is essentially no trend.

**Reply by robjhyndman** [reply]
```header
Originally posted by Rohit on 2017-06-29 08:59:37
````
Thanks for the simple explanation. Appreciate your time!

**Comment by robjhyndman** [comment]
```header
Originally posted by Pranav on 2017-09-08 19:45:04
````
It says the decomp method is not available. Can I know what is the
source library for decomp method used for decomposition?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-09-09 00:29:58
````
Read the post.

**Comment by robjhyndman** [comment]
```header
Originally posted by Prathamesh Gore on 2018-05-23 09:02:40
````
I have time series daily data of 570 days(\&lt;2 year). So I cant use
decompose function in R to find annual seasonality. Hence using other
similar time series I have found monthly seasonal indices and I have
deseasonalized my time series by dividing observed value by seasonality
index of that month.\
Now I want to find out trend &amp; randomness(error) in deseasonalised data.
How to use decompose function for deaseasonalised time series

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-05-23 21:36:15
````
This is not a help service. Ask on
[crossvalidated.com](http://crossvalidated.com &quot;crossvalidated.com&quot;)

**Comment by robjhyndman** [comment]
```header
Originally posted by ezgi suzgun on 2018-06-20 13:04:28
````
Thank you for posting the code Rob. I have a question about the code. I
am using another dataset with much smaller values than the gas data and
I am receiving an error which says &quot; Error in if (eax == Inf) f1eax \&lt;-
1 else f1eax \&lt;- (eax - 1)/(eax + b) :\
missing value where TRUE/FALSE needed &quot; in the skewness calculation on
adjusted data. Could it be because my dataset range? Should I play with
the a and b constants in the code? Thank you.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-06-20 13:53:03
````
I suggest you use the tsfeatures package (on github) instead. Presumably
eax is missing, so you will need to figure out why that is the case.

**Comment by robjhyndman** [comment]
```header
Originally posted by Marie on 2018-07-17 05:28:25
````
Hi. This is great, I need to do something similar (cluster sales people
based on their sales over time). Are you able to provide any pointers or
an-easy-to-decipher resource (book, etc) on how I could also quantify
&quot;shocks/interruptions&quot; (unsure of terminology here) to the time
series. I am using shocks here to refer to incentives provided to the
salesperson over different points in time to increase their sales. For
each sales person, I would like to determine values for trend,
seasonality, reaction to incentives, etc and then cluster them
accordingly. My time series knowledge is rusty, so any recommendations
of resources to get up to speed quickly would be also appreciated.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-07-17 05:30:49
````
Use the tsfeatures package:
[https://github.com/robjhynd\...](https://github.com/robjhyndman/tsfeatures)

**Reply by robjhyndman** [reply]
```header
Originally posted by Marie on 2018-08-28 04:11:23
````
Thank-you. I am looking into that package now and have since gone
through most of your book - Forecasting: Principles and Practice. It has
been a great help. Is this right - that there is some overlap (e.g. STL
decomposition, periodicity, nonlinearity) in the above functions and the
tsfeatures package, but there are some functions above (e.g. skewness,
self-similarity, chaos) not in tsfeatures and vice-versa (e.g.
linearity, acf decomposition). Would you therefore run both on the
original data, or are the features extracted in tsfeatures more robust/
generally sufficient in most application domains? Also, if using the
tsfeatures package, is it recommended I first perform Box-Cox
transformation (as above); run tsfeatures; back-transform using
InvBoxCox() and then use similar code to above to standardise to \[0,1\]
range for clustering? Thanks again for your help to date

**Reply by robjhyndman** [reply]
```header
Originally posted on 2018-08-28 06:41:50
````
Use the tsfeatures package, and add any features you want that are not
included. The tsfeatures package is designed to be extensible. It
depends on the data to whether you need to transform it. Whether you
standardise features depends on the subsequent analysis of the features.

**Comment by robjhyndman** [comment]
```header
Originally posted by Matty AB on 2019-01-31 04:28:33
````
Hi Rob. Thanks for sharing your code. I am dealing with a set of sales
time series that are impacted by promotion. I am testing a few
statistical and machine learning models on them and I want to see which
model works best in which conditions (I am interested in the variability
of time series caused with promotion and I use Coefficient of Variation
as the variability criteria).\
I see that you have used different characteristics but not Coefficient
of Variation. I was wondering if I can use CoV as one of time series
criteria to measure volatility? or is it better to rely on &quot;Lyapunov
Exponent&quot; and other criteria that you have used?\
Thanks in advance.

**Comment by robjhyndman** [comment]
```header
Originally posted by nand on 2019-02-14 07:17:40
````
Hi Professor. I have doubt regarding stl. I used stl to get the time
series components in a monthly series of 20 years. After adding
two/three data points to the series i did the same procedure, but got
different components(values of trend and seasonality). May i know if
there is anything where the history of trend/seasonality remain the same
(like X-12, X-13 Arima methods) even if we add new data points to the
series we already have. Thanks.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2019-02-14 07:19:56
````
Why should it be the same? You have more information when you add data.

**Comment by robjhyndman** [comment]
```header
Originally posted by Cristian Valdez on 2019-03-18 01:13:48
````
Thank you for the effort and dedication to your forecast package!

**Comment by robjhyndman** [comment]
```header
Originally posted by nand on 2019-03-19 06:44:25
````
Dear Prof. May i know any methods/packages to get the cycle/turn of a
time series.(like BBQ or any updated packages)

**Comment by robjhyndman** [comment]
```header
Originally posted by algorithmic on 2019-03-27 02:25:01
````
Is the findfrequency function in essence doing what the fourier
transform does?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2019-03-27 02:44:51
````
No. You could use a Fourier transform to try to estimate the dominant
frequency, but this function uses the spectrum of a fitted
autoregressive process instead.

**Comment by alanchn31** [comment]
Regarding decomposing the data into trend and seasonal components, is there a reason why you do an automated Box-Cox transformation before decomposing into trend and seasonal components? From my understanding, a log-transform can be used to transform the series so an additive decomposition can be done to it but I am not sure if there is a particular reason we should do Box-Cox transform?

**Reply by robjhyndman** [reply]
A Box-Cox is more general than a log transformation. Sometimes a log transformation is too strong, and then a weaker Box-Cox transformation can help to make the series better suited to an additive decomposition.
